### DISCUSSION TITLE
How to persist workflow state and resume particular activity from where it was Suspended

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#4928] [Design & tech]
As per ELSA workflow [documentation](https://v2.elsaworkflows.io/docs/installation/persistence) , there are different ways to persist workflow's state like EF core, Mongo DB, YesSQL etc. When to use which and their cons and pros? Which is better way for real time application. how to achieve using Mongo DB? After maintaining state how to resume particular activity from where it was Suspended? Can someone please explain with example?

**Comment by glime-ai** [comment]
To decide which persistence provider to use with ELSA Workflows, you should consider the specific needs of your application, such as performance requirements, existing infrastructure, developer expertise, and future scalability. Each persistence provider has its own set of pros and cons, and the best choice depends on the context of your application.

Here are some general considerations for each:

1. **Entity Framework Core (EF Core)**: This is a good choice if you are already using EF Core in your application and are familiar with it. It allows you to use SQL Server, SQLite, PostgreSQL, and other databases supported by EF Core. It's a mature ORM with good support and integration with .NET applications.

   - Pros: Well-integrated with .NET, supports LINQ queries, and has a wide range of database support.
   - Cons: Might be overkill for simple scenarios, and performance might not be as high as with some NoSQL databases for certain workloads.

2. **MongoDB**: This is a NoSQL database that is known for its scalability and performance with large amounts of data and high throughput applications.

   - Pros: Schema-less, which can be more flexible for certain types of applications. Good performance for large datasets.
   - Cons: NoSQL databases generally lack the transactional and relational capabilities of SQL databases, which can be a drawback for some applications.

3. **YesSQL**: This is a .NET document database interface that provides a bridge between the object world and the document world without requiring an ORM.

   - Pros: Simplifies the persistence of object graphs and can be a good fit for applications that don't need the full capabilities of a relational database.
   - Cons: Less common than other options, which might result in a steeper learning curve and less community support.

To achieve persistence using MongoDB, you would configure ELSA to use the MongoDB provider. Here's an example of how you might configure it:

```csharp
services
    .AddElsa(elsa => elsa
        .UseMongoDbPersistence(options => options.ConnectionString = "your-mongodb-connection-string")
        .AddConsoleActivities()
        .AddHttpActivities());
```

To resume a particular activity from where it was suspended, ELSA uses bookmarks. When a workflow is suspended, ELSA creates a bookmark for the activity that caused the suspension. When the event that the bookmark corresponds to occurs, ELSA uses the bookmark to resume the workflow from the point of suspension.

Here's a high-level example of how you might handle this:

```csharp
// Assuming you have a workflow instance that has been suspended
var workflowInstanceId = "your-workflow-instance-id";
var bookmarkName = "your-bookmark-name"; // The name of the bookmark to resume from
var bookmarkResult = await workflowRunner.RunWorkflowAsync(workflowInstanceId, bookmarkName);
```

Please note that the actual implementation will depend on the specifics of your application and the events you're handling.

For more detailed examples and documentation, you should refer to the official ELSA documentation and resources:

- ELSA Documentation: https://elsa-workflows.github.io/elsa-core/
- ELSA GitHub Repository: https://github.com/elsa-workflows/elsa-core
